src/backend_telemetry_buffer.rs

```diff
@@ -0,0 +1,169 @@
+    type BufferItem = composite_type!(COMPOSITE_TYPE_NAME);
+    struct Buffer {
+        mem_ctx: PgMemoryContexts,
+        list: PgList<BufferItem>,
+    }
+
+    struct BufferIter<'b> {
+        buffer: &'b mut Buffer,
+        pos: usize,
+    }
+
+    impl<'a> Iterator for BufferIter<'a> {
+        type Item = BufferItem;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            // We are not deallocaing anything here because Drop does it wholesale.
+            let item = self.buffer.list.get_ptr(self.pos);
+            self.pos += 1;
+            // SAFETY: push_rec is expected to be the only funciton appending to buffer.list
+            item.map(|i| unsafe {
+                PgHeapTuple::from_datum_in_memory_context(
+                    PgMemoryContexts::CurrentMemoryContext,
+                    i.into(),
+                    false,
+                    BufferItem::type_oid(),
+                )
+                .unwrap()
+            })
+        }
+
+        fn size_hint(&self) -> (usize, Option<usize>) {
+            let remaining = self.buffer.list.len() - self.pos;
+            (remaining, Some(remaining))
+        }
+    }
+
+    impl<'a> Drop for BufferIter<'a> {
+        fn drop(&mut self) {
+            // Drop on the list calls list_free(). For it to work without segfaults we first 
+            // need to replace the list (triggering the Drop) and only then call mem_ctx.reset()
+            self.buffer.list = PgMemoryContexts::CacheMemoryContext.switch_to(|_| PgList::new());
+            self.buffer.mem_ctx.reset();
+            self.buffer.list = self.buffer.mem_ctx.switch_to(|_| PgList::new());
+        }
+    }
+
+    fn init_buffer() -> &'static Mutex<Buffer> {
+        static mut SINGLETON: MaybeUninit<Mutex<Buffer>> = MaybeUninit::uninit();
+        static ONCE: Once = Once::new();
+
+        unsafe {
+            ONCE.call_once(|| {
+                let mut mem_ctx = PgMemoryContexts::CacheMemoryContext
+                    .switch_to(|_| PgMemoryContexts::new("backend_telemetry_buffer_context"));
+                let list = mem_ctx.switch_to(|_| PgList::new());
+                let singleton = Mutex::new(Buffer { mem_ctx, list });
+                SINGLETON.write(singleton);
+            });
+
+            SINGLETON.assume_init_ref()
+        }
+    }
+
+    #[pg_extern(volatile, strict, create_or_replace, requires = ["backend_telemetry_rec_decl"])]
+    pub fn push_rec(
+        _r: pgx::composite_type!(COMPOSITE_TYPE_NAME),
+        fcinfo: pg_sys::FunctionCallInfo,
+    ) -> bool {
+        let buffer_mutex = init_buffer();
+        let r_datum = pg_getarg_datum_raw(fcinfo, 0);
+        {
+            let mut q_guard = buffer_mutex.lock().unwrap();
+            let prev_ctx = q_guard.mem_ctx.set_as_current();
+            let detoasted = unsafe {
+                pg_sys::pg_detoast_datum_copy(r_datum.cast_mut_ptr() as *mut pg_sys::varlena)
+            };
+            q_guard.list.push(detoasted.cast());
```

Comment:
2 questions:
- shouldn't this go through a push function in Buffer. Not sure I love accessing the internal list element out here. 
- should we limit the size of the buffer to last X items. They way this is done, the list can grow indefinitely. I always imagined this was a ring buffer